File: Leetcode/1640. Check Array Formation Through Concatenation/solution.kt
class Solution {
    fun canFormArray(arr: IntArray, pieces: Array<IntArray>): Boolean {
        val index:HashMap<Int, IntArray> = HashMap<Int,IntArray>()
        
        for (elem in pieces) {
            index.put(elem[0], elem)
        }
        
        var result: List<Int> = listOf();
        
        for (num in arr) {
            val list:IntArray? = index.get(num)

            if(list?.isEmpty() == false) {
                result = result + list.toList()
            }
        }
        
        return result == arr.toList()
    }
}

File: Leetcode/1640. Check Array Formation Through Concatenation/solution.rs
use std::collections::HashMap;

impl Solution {
    pub fn can_form_array(arr: Vec<i32>, pieces: Vec<Vec<i32>>) -> bool {
        let mut index = HashMap::new();

        for elem in pieces {
            index.insert(elem[0], elem);
        }

        let mut result = Vec::new();

        for elem in &arr {
            match index.get(&elem) {
                Some(list) => {
                    for num in list {
                        result.push(*num);
                    }
                }
                None => continue,
            }
        }

        return result == arr;
    }
}
